quickwit/quickwit-indexing/src/source/doc_file_reader.rs

```diff
@@ -129,6 +135,142 @@ impl DocFileReader {
     }
 }
 
+#[async_trait]
+pub trait BatchReader: Send {
+    async fn read_batch(
+        &mut self,
+        source_ctx: &SourceContext,
+        source_type: SourceType,
+    ) -> anyhow::Result<BatchBuilder>;
+
+    fn is_eof(&self) -> bool;
+}
+
+pub struct ObjectUriBatchReader {
+    partition_id: PartitionId,
+    reader: DocFileReader,
+    current_offset: usize,
+    is_eof: bool,
+}
+
+impl ObjectUriBatchReader {
+    pub async fn try_new(
+        storage_resolver: &StorageResolver,
+        partition_id: PartitionId,
+        uri: &Uri,
+        position: Position,
+    ) -> anyhow::Result<Self> {
+        let current_offset = match position {
+            Position::Beginning => 0,
+            Position::Offset(offset) => offset
+                .as_usize()
+                .context("file offset should be stored as usize")?,
+            Position::Eof(_) => {
+                return Ok(ObjectUriBatchReader {
+                    partition_id,
+                    reader: DocFileReader::empty(),
+                    current_offset: 0,
+                    is_eof: true,
+                })
+            }
+        };
+        let reader = DocFileReader::from_uri(storage_resolver, uri, current_offset).await?;
+        Ok(ObjectUriBatchReader {
+            partition_id,
+            reader,
+            current_offset,
+            is_eof: false,
+        })
+    }
+}
+
+#[async_trait]
+impl BatchReader for ObjectUriBatchReader {
+    async fn read_batch(
+        &mut self,
+        source_ctx: &SourceContext,
+        source_type: SourceType,
+    ) -> anyhow::Result<BatchBuilder> {
+        let limit_num_bytes = self.current_offset + BATCH_NUM_BYTES_LIMIT as usize;
+        let mut new_offset = self.current_offset;
+        let mut batch_builder = BatchBuilder::new(source_type);
+        while new_offset < limit_num_bytes {
+            if let Some(record) = source_ctx.protect_future(self.reader.next_record()).await? {
+                new_offset = record.next_offset as usize;
+                batch_builder.add_doc(record.doc);
+            } else {
+                self.is_eof = true;
+                break;
+            }
+        }
+        if new_offset > self.current_offset {
+            let to_position = if self.is_eof {
+                Position::eof(new_offset)
+            } else {
+                Position::offset(new_offset)
+            };
+            batch_builder
+                .checkpoint_delta
+                .record_partition_delta(
+                    self.partition_id.clone(),
+                    Position::offset(self.current_offset),
+                    to_position,
+                )
+                .unwrap();
```

Comment:
unwrap?

Reply:
Took it from current main (https://github.com/quickwit-oss/quickwit/blob/84572ea5955ecbe2b0d57ae5a1fbe5d5600ce716/quickwit/quickwit-indexing/src/source/file_source.rs#L102-L109). Changine to a `?` instead.